src/Tizen.Applications.Cion/Tizen.Applications/ClientBase.cs

```diff
@@ -0,0 +1,363 @@
+        /// <since_tizen> 9 </since_tizen>
+        public void Disconnect()
+        {
+            Interop.Cion.ErrorCode ret = Interop.CionClient.CionClientDisconnect(_handle);
+            if (ret != Interop.Cion.ErrorCode.None)
+            {
+                throw CionErrorFactory.GetException(ret, "Failed to disconnect.");
+            }
+            _peer = null;
+        }
+
+        /// <summary>
+        /// Sends data synchronously to the connected cion server.
+        /// </summary>
+        /// <param name="data">The data to send.</param>
+        /// <param name="timeout">The timeout of sending operation.</param>
+        /// <exception cref="ArgumentException">Thrown when the given data is invalid.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when there is no connected cion server.</exception>
+        /// <since_tizen> 9 </since_tizen>
+        public byte[] SendData(byte[] data, int timeout)
+        {
+            Interop.Cion.ErrorCode ret = Interop.CionClient.CionClientSendData(_handle, data, data?.Length ?? -1, timeout, out byte[] returnData, out int returnDataSize);
+            if (ret != Interop.Cion.ErrorCode.None)
+            {
+                throw CionErrorFactory.GetException(ret, "Failed to send data.");
+            }
+
+            Log.Info(LogTag, string.Format("Returned data size: {0}", returnDataSize));
+
+            return returnData;
+        }
+
+        /// <summary>
+        /// Sends payload asynchronously to the connected cion server.
+        /// </summary>
+        /// <param name="payload">The payload to send.</param>
+        /// <exception cref="ArgumentException">Thrown when the payload is not valid.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when there is no connected cion server.</exception>
+        /// <since_tizen> 9 </since_tizen>
+        public Task<PayloadAsyncResult> SendPayloadAsync(Payload payload)
+        {
+            if (payload?.Id.Length == 0)
+            {
+                throw new ArgumentException("Payload is invalid.");
+            }
+
+            TaskCompletionSource<PayloadAsyncResult> tcs = new TaskCompletionSource<PayloadAsyncResult>();
+            _tcsDictionary[payload.Id] = tcs;
+
+            if (_payloadAsyncResultCb == null)
+            {
+                Interop.CionClient.CionClientPayloadAsyncResultCb cb = new Interop.CionClient.CionClientPayloadAsyncResultCb(
+                    (IntPtr result, IntPtr userData) =>
+                    {
+                        PayloadAsyncResult resultPayload = null;
+                        try
+                        {
+                            resultPayload = PayloadAsyncResult.CreateFromHandle(result);
+                        }
+                        catch (Exception e)
+                        {
+                            Log.Error(LogTag, string.Format("Failed to create PayloadAsyncResult from result handle: {0}.", e.Message));
```

Comment:
어쨌든 `tcsToReturn.SetException`으로 에러가 났음을 알리거나 SetResult에 빈 결과를 넣거나 하여 작업이 완료되었음을 알려야지 await로 대기중인 작업이 에러처리를 한든 해서 계속 진행될수 있을것 같습니다. 그렇지 않으면 await로 대기중인 상황에서 이미 끝난 Task에 계속 대기하는 상황이 발생할듯합니다.
```suggestion
                            Log.Error(LogTag, string.Format("Failed to create PayloadAsyncResult from result handle: {0}.", e.Message));
                            tcsToReturn.SetResult(new PayloadAsyncResult(PayloadAsyncResultCode.Error, null, null);
```
tcsToReturn.SetException

Reply:
애러상황에 SetException으로 애러처리를 할 수 있도록 수정하였습니다.